File: semester1/3-2/3-2/PriorityQueue.cpp
#include "pch.h"
#include "PriorityQueue.h"
#include <iostream>

#define _CRT_SECURE_NO_WARNINGS



struct Element
{
	int priority = -1;
	std::string data = " ";
	Element *next = nullptr;
};

void clear(PriorityQueue *&queue)
{
	if (queue->head != nullptr)
	{
		Element *&pointer = queue->head;
		Element *following = pointer->next;
		while (following != nullptr)
		{
			delete pointer;
			pointer = following;
			following = following->next;
		}
		delete pointer;
	}
}


void enqueue(PriorityQueue *&queue, const int priority, const std::string data)
{
	Element *&head = queue->head;
	auto *node = new Element{ priority, data, nullptr };
	if (head == nullptr)
	{
		head = node;
		return;
	}
	if (node->priority > head->priority)
	{
		node->next = head;
		head = node;
		return;
	}
	Element *pointer = head;
	while (pointer->next != nullptr && (pointer->next)->priority > node->priority)
	{
		pointer = pointer->next;
	}
	if (pointer->next == nullptr)
	{
		pointer->next = node;
		return;
	}
	if ((pointer->next)->priority <= node->priority)
	{
		node->next = pointer->next;
		pointer->next = node;
		return;
	}
}

std::string dequeue(PriorityQueue *&queue)
{
	Element *&head = queue->head;
	if (head == nullptr)
	{
		return "-1";
	}
	Element *temp = head;
	std::string toReturn = temp->data;
	head = head->next;
	delete temp;
	return toReturn;
}

File: semester2/T1-T1/T1-T1/PriorityQueue.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PriorutyQueueNamespace
{
    /// <summary>
    /// Queue, where element with higher priority goes earlier, then element with lower priority
    /// </summary>
    public class PriorityQueue
    {
        private Node head;
        public int Size { get; set; }
        public class Node
        {
            public Node Next { get; set; }
            public string Data { get; }
            public int Prior { get; }

            public Node(Node next, string data, int prior)
            {
                Next = next;
                Data = data;
                Prior = prior;
            }
        }

        /// <summary>
        /// Add an element in queue
        /// </summary>
        /// <param name="data">Data, you want to add to the queue</param>
        /// <param name="prior">Priority of your data</param>
        public void Enqueue(string data, int prior)
        {
            if (Size == 0)
            {
                head = new Node(null, data, prior);
                ++Size;
                return;
            }
            if (prior > head.Prior)
            {
                ++Size;
                head = new Node(head, data, prior);
            }
            var current = head;
            while (current.Next != null && prior >= current.Next.Prior)
            {
                current = current.Next;
            }
            current.Next = new Node(current.Next, data, prior);
            ++Size;
        }

        /// <summary>
        /// Returns data with highest priority
        /// </summary>
        /// <returns></returns>
        public string Dequeue()
        {
            if (Size == 0)
            {
                throw new RemovingElementOfEmptyQueue();
            }
            var toReturn = head.Data;
            head = head.Next;
            --Size;
            return toReturn;
        }

        /// <summary>
        /// Deletes all elements of your queue
        /// </summary>
        public void Clear()
        {
            Size = 0;
            head = null;
        }
    }
}
